### PR TITLE
[NativeAOT] Asynchronous thread suspension at GC safe points on Windows

### PR DESCRIPTION
**Pull request by VSadov**
Support for asynchronous suspension at GC safe points on Windows.

- ICodeManager::IsSafepoint API (Windows only, stubbed for Unix)
- asynchronous suspension via redirection of OS-suspended threads
- AVX support
- ARM64 support (not tested, not enabled)

Contributes to: #67805

### CODE REVIEW

File: src/coreclr/nativeaot/Runtime/thread.cpp [hunk 1]
```diff
@@ -949,20 +1026,11 @@ EXTERN_C NOINLINE void FASTCALL RhpWaitForSuspend2()
 // Standard calling convention variant and actual implementation for RhpWaitForGC
 EXTERN_C NOINLINE void FASTCALL RhpWaitForGC2(PInvokeTransitionFrame * pFrame)
 {
-
     Thread * pThread = pFrame->m_pThread;
-
     if (pThread->IsDoNotTriggerGcSet())
         return;
 
-    // The wait operation below may trash the last win32 error. We save the error here so that it can be
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.cpp:958]: Why did you remove the last error preservation here?

Comment (jkotas) [reply src/coreclr/nativeaot/Runtime/thread.cpp:958]: Ah ok, it was moved to WaitForGC.

File: src/coreclr/nativeaot/Runtime/thread.cpp [hunk 2]
```diff
@@ -769,6 +801,51 @@ bool Thread::InternalHijack(PAL_LIMITED_CONTEXT * pSuspendCtx, void * pvHijackTa
...
+    CONTEXT* redirectionContext = GetRedirectionContext();
+    if (redirectionContext == NULL)
+        return false;
+
+    StackFrameIterator frameIterator(this, pSuspendCtx);
+
+    if (!frameIterator.IsValid())
+        return false;
+
+    frameIterator.CalculateCurrentMethodState();
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.cpp:829]: Why call `CalculateCurrentMethodState`? `frameIterator` is not used after this point.

File: src/coreclr/nativeaot/Runtime/thread.cpp [hunk 3]
```diff
@@ -974,6 +1042,65 @@ EXTERN_C NOINLINE void FASTCALL RhpGcPoll2(PInvokeTransitionFrame* pFrame)
...
+    memcpy(&pPreservedRegs[11], &pCtx->X0, sizeof(uint64_t) * 19);
+    // LR
+    pPreservedRegs[30] = pCtx->Lr;
+#elif
+    ASSERT_UNCONDITIONALLY("NYI for this arch");
+#endif
+
+    pThread->WaitForGC(frame);
+
+    // restore execution at interrupted location
```
Comment (jkotas) [inline src/coreclr/nativeaot/Runtime/thread.cpp:1097]: Do we need to propagate the register values back in case the GC updated them?

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.cpp:1097]: Right. Forgot about that part. :)

Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.cpp:1097]: I am also thinking about making it possible to seed stackwalks with CONTEXT* on Windows.
That may allow avoiding copying the registers here and in Pal, and remove alloca. 
Probably not in this PR though, since I want to unblock libraries testing.

File: src/coreclr/nativeaot/Runtime/thread.cpp (hunk 2 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.cpp:829]: It was copied from `InternalHijack` initially. I now think the entire `frameIterator` deal is unnecessary here.

File: src/coreclr/nativeaot/Runtime/thread.cpp (hunk 3 above)
Comment (VSadov) [reply src/coreclr/nativeaot/Runtime/thread.cpp:1097]: Actually not copying the registers solves both problems - no redundant copying and no need to reflect back the changes.

**Comment by VSadov** [comment]
Hmm. tests seem to pass locally.

**Comment by VSadov** [comment]
It was quite reproducible and simple and library tests catch the issues (I only ran smoke tests locally).

GC decoder expects an off-by-one skew for nonleaf frames and we used to apply the skew unconditionally. Now that we have leaf frames to scan, we need to be more intentional with that.

**Comment by VSadov** [comment]
> You may want to think about how to stress test this before it gets merged.

Running library tests in a loop seems to be a good stress for this feature. The tests themselves and xunit present all kinds of situations for stackwalks.

I have run the entire set 30 times in Release and Debug each. 
I do not see any new failures, but there is an occasional assert that happens every 5 runs or so with or without this change. The typical scenario is when we have a lot of threads running and one throws cancellation exception, while another thread starts GC suspension. 

Then we hit an assert in:

https://github.com/dotnet/runtime/blob/b44fd4aeffafd68bc710b2feec66318835ca3edc/src/coreclr/nativeaot/Runtime/thread.cpp#L66

```
  Message: Not allowed when suspended for GC.
  Expression: 'ThreadStore::GetSuspendingThread() == NULL'


[0x0]   KERNELBASE!RaiseFailFastException + 0x152   
[0x1]   System_Runtime_Tests!PalRaiseFailFastException + 0x28   
[0x2]   System_Runtime_Tests!Assert + 0x156   
[0x3]   System_Runtime_Tests!Thread::GetTransitionFrameForStackTrace + 0x34   
[0x4]   System_Runtime_Tests!StackFrameIterator::InternalInitForStackTrace + 0x4d   
[0x5]   System_Runtime_Tests!RhpSfiInit + 0x3a   
[0x6]   System_Runtime_Tests!S_P_CoreLib_System_Runtime_RuntimeExports__RhpCalculateStackTraceWorker + 0x48   
[0x7]   System_Runtime_Tests!RhpGetCurrentThreadStackTrace + 0x33   
[0x8]   System_Runtime_Tests!S_P_CoreLib_System_Runtime_RuntimeExports__RhGetCurrentThreadStackTrace + 0x90   
[0x9]   System_Runtime_Tests!S_P_CoreLib_System_Diagnostics_StackTrace__InitializeForCurrentThread + 0x20   
[0xa]   System_Runtime_Tests!S_P_CoreLib_System_Diagnostics_StackTrace___ctor_0 + 0xf   
[0xb]   System_Runtime_Tests!S_P_CoreLib_System_Exception__SetCurrentStackTrace + 0x54   
```

I think the assert is bogus. We are in native code in coop mode, we certainly can scan our own stack. Besides, suspension may happen right after we check anyways.

I am going to remove the assert. Otherwise, I think this PR is good to merge.

**Comment by VSadov** [comment]
Also `CreateThreadLocalContentionCountObject` can be reentrant on NativeAOT. Creating a sub-counter takes locks. If we contend again, we will assert and deadlock. 

This is fairly rare (I saw it just once) and completely unrelated to this change since it is entirely in managed code. 
I will log an issue to follow up with that separately.

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by VSadov** [comment]
Thanks!!

**Comment by VSadov** [comment]
@LakshanF hangs in GC suspension should be fixed on Windows x64 now.
